asim: use queues with LBRebalancingMultiMetricOnly

wenyihu6 opened the pull request:
Epic: none
Release note: none

---
**asim: use queues with LBRebalancingMultiMetricOnly**

This commit updates the mma-only mode to enable replicate and lease queues while
using LBRebalancingMultiMetricOnly. In this mode, constraint repairs in the
replicate and lease queues should still occur with mma enabled, but range
count–based rebalancing is disabled.

It’s a bit surprising that this diff didn’t cause any datadriven test changes. I
think this is likely because the mma tests currently don’t exercise constraint
repair. Tests that do cover constraint repair and lease preference likely don’t
run in mma-only mode. From a range/lease count rebalancing perspective, this
makes sense, as this mode previously disabled both replicate and lease queues.
Future commits will add test coverage to verify that this mode supports constraint
repair with mma.

---
**asim: rename mode from both to mma-and-count**

This commit renames the mode from both to mma-and-count to clarify that it
enables the replica and lease count based rebalancing.

---
**asim: schedule event with default**

Previously, the default configuration did not schedule an event. This would
cause issues in cases like cfg(mma-only,default), where mma-only would schedule
an event to change the cluster setting, but default would not reset it in the
next round, leaving the cluster setting incorrectly persisted. This commit fixes
the issue by ensuring that default also sets the cluster setting as an event.
The persistence of settings and cluster settings across simulations is a bit
troublesome. For now, this fix is sufficient since we only allow rebalance_mode
to be a delay event. We may need to revisit this in the future.

---
**asim: add mma-only mode to some existing asim tests**

This commit adds mma-only mode to some existing asim tests that exercise
constraint conformance to demonstrate that mma-only mode can pass assertions
with the replicate and lease queues enabled. Note that some assertions on
QPS/lease fail which worth investigating further. I took a quick pass - CPU
usage was balanced, but QPS was not.

Note that this commit removes the default mode for these tests so that changes
can be compared more clearly with git diff, avoiding side-by-side comparisons
between two modes. A follow-up commit will reintroduce the default mode
alongside mma-only.

---
**asim: add the default mode back for asim tests**

This commit adds the default mode for some asim tests.

---
**kvserverbase: update validation check for LoadBasedRebalancingMode**

Previously, we introduced LBRebalancingMultiMetricOnly and renamed
LBRebalancingMultiMetric to LBRebalancingMultiMetricAndCount. However, a
hardcoded enum string check in the cluster setting was overlooked and not
updated. This commit fixes that by updating the check correctly.

wenyihu6 committed:
asim: use queues with LBRebalancingMultiMetricOnly

This commit updates the mma-only mode to enable replicate and lease queues while
using LBRebalancingMultiMetricOnly. In this mode, constraint repairs in the
replicate and lease queues should still occur with mma enabled, but range
count–based rebalancing is disabled.

It’s a bit surprising that this diff didn’t cause any datadriven test changes. I
think this is likely because the mma tests currently don’t exercise constraint
repair. Tests that do cover constraint repair and lease preference likely don’t
run in mma-only mode. From a range/lease count rebalancing perspective, this
makes sense, as this mode previously disabled both replicate and lease queues.
Future commits will add test coverage to verify that this mode supports constraint
repair with mma.

wenyihu6 committed:
asim: rename mode from both to mma-and-count

This commit renames the mode from both to mma-and-count to clarify that it
enables the replica and lease count based rebalancing.

wenyihu6 committed:
asim: add mma-only mode to some existing asim tests

This commit adds mma-only mode to some existing asim tests that exercise
constraint conformance to demonstrate that mma-only mode can pass assertions
with the replicate and lease queues enabled. Note that some assertions on
QPS/lease fail which worth investigating further. I took a quick pass - CPU
usage was balanced, but QPS was not.

Note that this commit removes the default mode for these tests so that changes
can be compared more clearly with git diff, avoiding side-by-side comparisons
between two modes. A follow-up commit will reintroduce the default mode
alongside mma-only.

wenyihu6 committed:
asim: schedule event with default

Previously, the default configuration did not schedule an event. This would
cause issues in cases like cfg(mma-only,default), where mma-only would schedule
an event to change the cluster setting, but default would not reset it in the
next round, leaving the cluster setting incorrectly persisted. This commit fixes
the issue by ensuring that default also sets the cluster setting as an event.
The persistence of settings and cluster settings across simulations is a bit
troublesome. For now, this fix is sufficient since we only allow rebalance_mode
to be a delay event. We may need to revisit this in the future.

wenyihu6 committed:
asim: add the default mode back for asim tests

This commit adds the default mode for some asim tests.

wenyihu6 committed:
kvserverbase: update validation check for LoadBasedRebalancingMode

Previously, we introduced LBRebalancingMultiMetricOnly and renamed
LBRebalancingMultiMetric to LBRebalancingMultiMetricAndCount. However, a
hardcoded enum string check in the cluster setting was overlooked and not
updated. This commit fixes that by updating the check correctly.

tbg reviewed (approved):
> @tbg reviewed 1 of 1 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 18 of 18 files at r4, 18 of 18 files at r5, 5 of 5 files at r6, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153484)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 33 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153484#-O_GvpCg-32IkwcG2jf6:-O_GvpCg-32IkwcG2jf7:bqeqbkp):*
Could we give each instance a cloned settingsGen so that they truly start off from the same settings?
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153484#-O_Gx0fy1AkgPYwl9kQg:-O_Gx0fy1AkgPYwl9kQh:brluv6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/704bcbd904232241df9e28faf578b30d592f9829/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L60)):*
> ```help
> 	max/mean=4.00 tick=3
> 	max/mean=4.00 tick=4
> 	max/mean=4.00 tick=5
> ```

I filed and sheeted https://github.com/cockroachdb/cockroach/issues/153565 to track addressing these failures.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/decommission_conformance.txt` line 29 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/153484#-O_GxB8XBGxvn2ulfBr2:-O_GxB8XBGxvn2ulfBr3:b7bombu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f058aabb29aeeb63ab19eb88768c5c0fba8bce0/pkg/kv/kvserver/asim/tests/testdata/non_rand/decommission_conformance.txt#L29)):*
> ```help
> ----
> 
> eval duration=20m cfgs=(default,mma-only)
> ```

Remind me, how is `default` different from `sma-only`? Maybe we can remove `default`; it's not going to be obvious to most readers what is enabled in this mode anyway.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
asim: rename sma-only to sma-count and remove default

This commit renames sma-only to sma-count specify that it enables single-metric
store rebalancer and count-based rebalancing. It also removes default since it
is unclear what it was intended for and does the same thing as sma-count now.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153484)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`-- commits` line 33 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153484#-O_GvpCg-32IkwcG2jf6:-O_IDk__1c-A1cwvXwFg:bttiz0k):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could we give each instance a cloned settingsGen so that they truly start off from the same settings?
</blockquote></details>

I’ll add a new item to the spreadsheet. I wasn’t able to find a clean way to handle this - maybe you have a smart idea. For this `settingGen`, we essentially need a copy of the setting just before the first eval starts, since scheduled events will then begin modifying it. Making a real deep copy is tricky because the cluster setting field in the settings is a pointer and can’t be easily deep copied (is there an easy way to deep copy cluster settings?).

One idea was to reset the cluster setting to its default before each simulation run, but any changes that occurred before the first eval are part of the test setup and need to be preserved. This gets trickier when there are two evaluations.

I see two possible approaches:
1.  Introduce a revert function for scheduled events that modify the cluster settings.
2.  Enforce that all setting changes must be scheduled events (with `delay=0` if immediate) and we just use a new DefaultSimulationSettings every simulation run. In this case, changing the rebalance mode to default still needs to be a scheduled event. 

I think I like the second approach more wdyt? 
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/decommission_conformance.txt` line 29 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/153484#-O_GxB8XBGxvn2ulfBr2:-O_IF6Sk9wfHPGyH-FnZ:bx36way) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f058aabb29aeeb63ab19eb88768c5c0fba8bce0/pkg/kv/kvserver/asim/tests/testdata/non_rand/decommission_conformance.txt#L29)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Remind me, how is `default` different from `sma-only`? Maybe we can remove `default`; it's not going to be obvious to most readers what is enabled in this mode anyway.
</blockquote></details>

I’m confuse too - they seem to do the same thing according to `datadriven_simulation_test.go`. Just a guess - perhaps the original intention for `sma-only` was to include just the store rebalancer without the replicate/lease queue, since at the time we only had `mma-only` mode without the replicate/lease queue, which gave a more direct comparison? Anyway, now that `mma-only` has both replicate and lease queue enabled, we shouldn't need to disable the replicate/lease queue for other modes.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153484#-O_Gx0fy1AkgPYwl9kQg:-O_IFq1U-RZeWT7kfRO0:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/704bcbd904232241df9e28faf578b30d592f9829/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L60)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I filed and sheeted https://github.com/cockroachdb/cockroach/issues/153565 to track addressing these failures.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->
